lib/stages/FEngine.cpp

```diff
@@ -0,0 +1,545 @@
+    dish_separation_y(config.get<float>(unique_name, "dish_separation_y")),
+    num_dishes(config.get<int>(unique_name, "num_dishes")),
+    dish_locations(config.get<std::vector<int>>(unique_name, "dish_locations")),
+    // ADC
+    adc_frequency(config.get<float>(unique_name, "adc_frequency")),
+    num_taps(config.get<int>(unique_name, "num_taps")),
+    num_frequencies(config.get<int>(unique_name, "num_frequencies")),
+    num_times(config.get<int>(unique_name, "num_times")),
+    // Baseband beamformer setup
+    bb_num_dishes_M(config.get<int>(unique_name, "bb_num_dishes_M")),
+    bb_num_dishes_N(config.get<int>(unique_name, "bb_num_dishes_N")),
+    bb_num_beams_P(config.get<int>(unique_name, "bb_num_beams_P")),
+    bb_num_beams_Q(config.get<int>(unique_name, "bb_num_beams_Q")),
+    bb_beam_separation_x(config.get<float>(unique_name, "bb_beam_separation_x")),
+    bb_beam_separation_y(config.get<float>(unique_name, "bb_beam_separation_y")),
+    bb_num_beams(bb_num_beams_P * bb_num_beams_Q),
+    // Upchannelizer setup
+    upchannelization_factor(config.get<int>(unique_name, "upchannelization_factor")),
+    // FRB beamformer setup
+    // Pipeline
+    num_frames(config.get<int>(unique_name, "num_frames")),
+    // Frame sizes
+    E_frame_size(std::int64_t(1) * num_dishes * num_polarizations * num_frequencies * num_times),
+    A_frame_size(std::int64_t(1) * num_components * num_dishes * bb_num_beams * num_polarizations
+                 * num_frequencies),
+    J_frame_size(std::int64_t(1) * num_times * num_polarizations * num_frequencies * bb_num_beams),
+    S_frame_size(std::int64_t(1) * sizeof(short) * 2 * num_dish_locations),
+    G_frame_size(std::int64_t(1) * sizeof(float16_t) * num_frequencies * upchannelization_factor),
+    W_frame_size(std::int64_t(1) * sizeof(float16_t) * num_components * num_dish_locations_M
+                 * num_dish_locations_N * num_polarizations * num_frequencies),
+    // Buffers
+    E_buffer(get_buffer("E_buffer")), A_buffer(get_buffer("A_buffer")),
+    J_buffer(get_buffer("J_buffer")), S_buffer(get_buffer("S_buffer")),
+    G_buffer(get_buffer("G_buffer")), W_buffer(get_buffer("W_buffer")) {
+    assert(num_dishes <= num_dish_locations);
+    assert(std::ptrdiff_t(dish_locations.size()) == 2 * num_dish_locations);
+
+    assert(E_buffer);
+    assert(A_buffer);
+    assert(J_buffer);
+    assert(S_buffer);
+    assert(G_buffer);
+    assert(W_buffer);
+    register_producer(E_buffer, unique_name.c_str());
+    register_producer(A_buffer, unique_name.c_str());
+    register_producer(J_buffer, unique_name.c_str());
+    register_producer(S_buffer, unique_name.c_str());
+    register_producer(G_buffer, unique_name.c_str());
+    register_producer(W_buffer, unique_name.c_str());
+
+    INFO("Starting Julia...");
+    kotekan::juliaStartup();
+
+    INFO("Defining Julia code...");
+    {
+        std::ifstream file("lib/stages/FEngine.jl");
```

Comment:
Could you please add some error checking here, eg if the file doesn't exist (eg, because the user isn't running kotekan from the correct directory)?  From experience, I can say that the resulting error message is not very intuitive!

Reply:
Done.